mlx/ops.cpp

```diff
@@ -2758,4 +2760,91 @@ array dequantize(
   return w_full;
 }
 
+std::map<char, int> string_to_char_map(std::string inp) {
+  std::map<char, int> counts;
+  for (int i = 0; i < inp.size(); i++) {
+    counts[inp[i]] = i;
+  }
+  return counts;
+}
+
+array einsum(
+    std::string equation,
+    const std::vector<array>& operands,
+    StreamOrDevice s /* = {} */) {
+  std::string output;
+  std::string input = equation;
```

Comment:
I would just initialize empty and set in both branches, rather than copy once and then copy again in one branch below. (IDK maybe there is copy on write but maybe not!)